### PR TITLE
Wp 1410 auth method

### PR DESCRIPTION
**Pull request by pc-m**
Depends-On: https://github.com/wazo-platform/wazo-auth-client/pull/50

### CODE REVIEW

File: wazo_auth/services/idp.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+from wazo_auth.services.helpers import BaseService
+
+
+class IDPService(BaseService):
+    def add_user(self, idp_type, user_uuid):
+        user = self._dao.user.get(user_uuid)
+        user.authentication_method = idp_type
+        self._dao.user.session.flush()
```
Comment (DrPyser) [inline wazo_auth/services/idp.py:11]: I disapprove. 
mutation of database objects should go through a proper DAO interface.

Also, `flush` does not commit the database transaction. When does the `commit` occurs? Using a DAO method, you should be able to expect the method call to represent a self-contained database transaction.

File: wazo_auth/services/idp.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+        return user
+
+    def remove_user(self, idp_type, user_uuid):
+        user = self._dao.user.get(user_uuid)
+        if user.authentication_method == idp_type:
+            user.authentication_method = 'default'
+            self._dao.user.session.flush()
```
Comment (DrPyser) [inline wazo_auth/services/idp.py:18]: see above

File: wazo_auth/services/idp.py (hunk 1 above)
Comment (pc-m) [reply wazo_auth/services/idp.py]: Fixed

File: wazo_auth/services/idp.py (hunk 2 above)
Comment (pc-m) [reply wazo_auth/services/idp.py]: Fixed

**Comment by DrPyser** [comment]
@pc-m can you check the code duplication warning from sonarcloud?

**Comment by pc-m** [comment]
> @pc-m can you check the code duplication warning from sonarcloud?

done

**Comment by pc-m** [comment]
recheck

File: wazo_auth/services/authentication.py [hunk 3]
```diff
@@ -1,36 +1,109 @@
...
+
+    def verify_saml(self, saml_session_id):
+        logger.debug('verifying SAML login')
+        saml_login = self._saml_service.get_user_login_and_remove_context(
+            saml_session_id,
+        )
+        if not saml_login:
+            raise NoMatchingSAMLSession(saml_session_id)
+
+        login = saml_login[0]
```
[the remark this answers is not in the document]
Comment (pc-m) [reply wazo_auth/services/authentication.py]: the saml AVA is a list

File: wazo_auth/services/authentication.py [hunk 4]
```diff
@@ -1,36 +1,109 @@
...
+        if (
+            authorized_authentication_method := self._authorized_authentication_method(
+                login
+            )
+        ) != 'saml':
+            raise UnauthorizedAuthenticationMethod(authorized_authentication_method)
+
+        # There's no SAML backend
+        backend = backend = self._get_backend('wazo_user')
```
[the remark this answers is not in the document]
Comment (pc-m) [reply wazo_auth/services/authentication.py]: We are using the `get_acl` and `get_metadata` of the wazo_user backend for the properties of the generated token

[the remark this answers is not in the document]
Comment (pc-m) [reply wazo_auth/services/authentication.py]: We do not use the verify_password function but everything else is required

**Review by DrPyser — DISMISSED** [review]

**Comment by DrPyser** [comment]
regate

**Comment by DrPyser** [comment]
regate

**Review by DrPyser — APPROVED** [review]

**Pull request merged**